SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.cpp

```diff
@@ -2,5 +2,130 @@
 
 #include "Schema/UnrealObjectRef.h"
 
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Utils/SchemaUtils.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogUnrealObjectRef, Log, All);
+
 const FUnrealObjectRef FUnrealObjectRef::NULL_OBJECT_REF = FUnrealObjectRef(0, 0);
 const FUnrealObjectRef FUnrealObjectRef::UNRESOLVED_OBJECT_REF = FUnrealObjectRef(0, 1);
+
+UObject* FUnrealObjectRef::ToObjectPtr(const FUnrealObjectRef& ObjectRef, USpatialPackageMapClient* PackageMap, bool& bOutUnresolved)
+{
+	if (ObjectRef == FUnrealObjectRef::NULL_OBJECT_REF)
+	{
+		return nullptr;
+	}
+	else
+	{
+		if (ObjectRef.bSingletonRef)
+		{
+			// This is a singleton ref, which means it's just the UnrealObjectRef of the singleton class, with this boolean set.
+			// Unset it to get the original singleton class UnrealObjectRef, and look it up in the PackageMap.
+			FUnrealObjectRef SingletonClassRef = ObjectRef;
+			SingletonClassRef.bSingletonRef = false;
+
+			return PackageMap->GetSingletonByClassRef(SingletonClassRef);
+		}
+
+		FNetworkGUID NetGUID = PackageMap->GetNetGUIDFromUnrealObjectRef(ObjectRef);
+		if (NetGUID.IsValid())
+		{
+			UObject* Value = PackageMap->GetObjectFromNetGUID(NetGUID, true);
+			if (Value == nullptr)
+			{
+				// At this point, we're unable to resolve a stably-named actor by path. This likely means either the actor doesn't exist, or
+				// it's part of a streaming level that hasn't been streamed in. Native Unreal networking sets reference to nullptr and continues.
+				// So we do the same.
+				FString FullPath;
+				SpatialGDK::GetFullPathFromUnrealObjectReference(ObjectRef, FullPath);
+				UE_LOG(LogUnrealObjectRef, Verbose, TEXT("Object ref did not map to valid object. Streaming level not loaded or actor deleted. Will be set to nullptr: %s %s"),
```

Comment:
Realise this isn't your change, but I feel with level streaming functionality we have in now, this message should be a warning, not verbose.